### PR TITLE
Add diagnostics to discovery announcer

### PR DESCRIPTION
**Pull request by dain**
Sometimes there are gaps in the discovery announcment, so add logging to
determine where gaps are happening.

### CODE REVIEW

File: discovery/src/main/java/io/airlift/discovery/client/Announcer.java [hunk 1]
```diff
@@ -23,8 +23,10 @@
 import com.google.common.util.concurrent.Futures;
```
Comment (nileema) [inline discovery/src/main/java/io/airlift/discovery/client/Announcer.java]: Fix typo in commit message JXM -> JMX

File: discovery/src/main/java/io/airlift/discovery/client/Announcer.java [hunk 2]
```diff
@@ -125,26 +126,36 @@ public void removeServiceAnnouncement(UUID serviceId)
...
-            public void onSuccess(Duration duration)
+            public void onSuccess(Duration expectedDelay)
             {
                 errorBackOff.success();
 
                 // wait 80% of the suggested delay
-                duration = new Duration(duration.toMillis() * 0.8, MILLISECONDS);
-                scheduleNextAnnouncement(duration);
+                expectedDelay = new Duration(expectedDelay.toMillis() * 0.8, MILLISECONDS);
+                log.debug("Service announcement succeeded after %s.  Next request will happen within %s", Duration.nanosSince(requestStart), expectedDelay);
```
Comment (electrum) [inline discovery/src/main/java/io/airlift/discovery/client/Announcer.java]: Use one space in messages

File: discovery/src/main/java/io/airlift/discovery/client/Announcer.java [hunk 3]
```diff
@@ -125,26 +126,36 @@ public void removeServiceAnnouncement(UUID serviceId)
...
+
+                scheduleNextAnnouncement(expectedDelay);
             }
 
             @Override
             public void onFailure(Throwable t)
             {
                 Duration duration = errorBackOff.failed(t);
+                // todo this is a duplicate log message and should be remove after root cause of announcement delay is determined
+                log.error("Service announcement failed after %s.  Next request will happen within %s", Duration.nanosSince(requestStart), expectedDelay);
```
Comment (electrum) [inline discovery/src/main/java/io/airlift/discovery/client/Announcer.java]: Same

File: discovery/src/main/java/io/airlift/discovery/client/Announcer.java [hunk 4]
```diff
@@ -69,6 +73,13 @@ public Announcer(DiscoveryAnnouncementClient announcementClient, Set<ServiceAnno
         this.announcementClient = announcementClient;
         serviceAnnouncements.forEach(this::addServiceAnnouncement);
         executor = new ScheduledThreadPoolExecutor(5, daemonThreadsNamed("Announcer-%s"));
+        executorMBean = new ThreadPoolExecutorMBean((ThreadPoolExecutor) executor);
+    }
+
+    @Managed
+    public ThreadPoolExecutorMBean getExecutorMBean()
```
Comment (electrum) [inline discovery/src/main/java/io/airlift/discovery/client/Announcer.java]: I think this should be `getExecutor()` so the name comes out correctly in JMX.

Does this need `@Nested`? All the ones in Presto have that.

**Comment by electrum** [comment]
Looks good

**Comment by nileema** [comment]
Assuming travis passes, looks good.

File: discovery/src/main/java/io/airlift/discovery/client/Announcer.java [hunk 5]
```diff
@@ -67,20 +71,25 @@ public Announcer(DiscoveryAnnouncementClient announcementClient, Set<ServiceAnno
...
         Preconditions.checkState(!executor.isShutdown(), "Announcer has been destroyed");
         if (started.compareAndSet(false, true)) {
             // announce immediately, if discovery is running
-            ListenableFuture<Duration> announce = announce();
+            ListenableFuture<Duration> announce = announce(System.nanoTime(), new Duration(0, SECONDS));
             try {
-                announce.get(30, TimeUnit.SECONDS);
+                announce.get(30, SECONDS);
             }
             catch (Exception ignored) {
```
Comment (nileema) [inline discovery/src/main/java/io/airlift/discovery/client/Announcer.java]: Log if we get timeoutException?

Comment (dain) [inline discovery/src/main/java/io/airlift/discovery/client/Announcer.java]: This is just the start method for the announcer

File: discovery/src/main/java/io/airlift/discovery/client/Announcer.java (hunk 4 above)
Comment (dain) [inline discovery/src/main/java/io/airlift/discovery/client/Announcer.java]: Yes to both

**Pull request merged**